File: apps/web/src/api/audit.ts
import { apiRequest } from "./http";
import type { ResourceMutationContext } from "./resources";

export interface ServerAuditEntity {
  entityType: string;
  entityId: string;
}

export interface ServerAuditEvent {
  id: string;
  actorType: "user" | "agent" | "system";
  actorId?: string;
  action: string;
  requestId: string;
  beforeData?: unknown;
  afterData?: unknown;
  metadata: Record<string, unknown>;
  entities: ServerAuditEntity[];
  createdAt: string;
  undoable: boolean;
}

export interface AuditEventPage {
  events: ServerAuditEvent[];
  nextCursor?: string;
  hasMore: boolean;
}

export interface UndoAuditResult {
  originalAuditId: string;
  entityType: string;
  entityId: string;
  entityOperation: string;
  entityVersion: number;
  data?: unknown;
}

function mutationHeaders(context: ResourceMutationContext, version: number): HeadersInit {
  return {
    "X-Device-ID": context.deviceId,
    "Idempotency-Key": context.mutationId,
    "If-Match": `"${version}"`,
  };
}

export function listAuditEvents(options: { cursor?: string; limit?: number } = {}): Promise<AuditEventPage> {
  const query = new URLSearchParams();
  if (options.cursor) query.set("cursor", options.cursor);
  if (options.limit !== undefined) query.set("limit", String(options.limit));
  const encoded = query.toString();
  return apiRequest(`/audit-events${encoded ? `?${encoded}` : ""}`);
}

export function getAuditEvent(auditEventId: string): Promise<ServerAuditEvent> {
  return apiRequest(`/audit-events/${auditEventId}`);
}

export function undoAuditEvent(auditEventId: string, version: number, context: ResourceMutationContext): Promise<UndoAuditResult> {
  return apiRequest(`/audit-events/${auditEventId}/undo`, { method: "POST", headers: mutationHeaders(context, version) });
}

export function auditEntityVersion(event: { afterData?: unknown }): number | undefined {
  if (!event.afterData || typeof event.afterData !== "object" || Array.isArray(event.afterData)) return undefined;
  const version = (event.afterData as { version?: unknown }).version;
  return typeof version === "number" && Number.isSafeInteger(version) && version > 0 ? version : undefined;
}

File: apps/api/internal/model/audit.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type AuditEvent struct {
	ID         uuid.UUID       `json:"id"`
	ActorType  string          `json:"actorType"`
	ActorID    *uuid.UUID      `json:"actorId,omitempty"`
	Action     string          `json:"action"`
	RequestID  uuid.UUID       `json:"requestId"`
	BeforeData json.RawMessage `json:"beforeData,omitempty"`
	AfterData  json.RawMessage `json:"afterData,omitempty"`
	Metadata   json.RawMessage `json:"metadata"`
	Entities   []AuditEntity   `json:"entities"`
	CreatedAt  time.Time       `json:"createdAt"`
	Undoable   bool            `json:"undoable"`
}

type UndoResult struct {
	OriginalAuditID uuid.UUID       `json:"originalAuditId"`
	EntityType      string          `json:"entityType"`
	EntityID        uuid.UUID       `json:"entityId"`
	EntityOperation string          `json:"entityOperation"`
	EntityVersion   int64           `json:"entityVersion"`
	Data            json.RawMessage `json:"data,omitempty"`
	BeforeData      json.RawMessage `json:"-"`
	AfterData       json.RawMessage `json:"-"`
}
